sql/spanner/inspect.go

```diff
@@ -0,0 +1,738 @@
+	var typ schema.Type
+
+	col := &columnDesc{}
+	spannerType = strings.TrimSpace(strings.ToUpper(spannerType))
+
+	// Split up type into, base type, size, and other modifiers.
+	re := regexp.MustCompile(`(\w+)(?:\((-?\d+|MAX)\))?`)
+	m := removeEmptyStrings(re.FindStringSubmatch(spannerType))
+	if len(m) == 0 {
+		return nil, fmt.Errorf("columnType: Invalid type: %q", spannerType)
+	}
+	col.typ = m[1]
+
+	if len(m) > 2 {
+		if m[2] == "MAX" {
+			col.sizeIsMax = true
+		} else {
+			size, err := strconv.Atoi(m[2])
+			if err != nil {
+				return nil, fmt.Errorf("columnType: Unable to convert %q to an int: %w", m[2], err)
+			}
+			col.size = size
+		}
+	}
+
+	switch col.typ {
+	case TypeInt64:
+		typ = &schema.IntegerType{T: col.typ}
+	case TypeBool:
+		typ = &schema.BoolType{T: col.typ}
+	case TypeTimestamp:
+		typ = &schema.TimeType{T: col.typ}
+	case TypeDate:
+		typ = &schema.TimeType{T: col.typ}
+	case TypeJSON:
+		typ = &schema.JSONType{T: col.typ}
+	case TypeNumeric:
+		typ = &schema.DecimalType{T: col.typ}
+	default:
+		if strings.HasPrefix(col.typ, TypeString) {
+			typ = &schema.StringType{
+				T:    col.typ,
+				Size: col.size,
+				Attrs: []schema.Attr{
+					&SizeIsMax{Bool: col.sizeIsMax},
+				},
+			}
+		} else if strings.HasPrefix(col.typ, TypeBytes) {
+			typ = &schema.BinaryType{
+				T:    col.typ,
+				Size: &col.size,
+				Attrs: []schema.Attr{
+					&SizeIsMax{Bool: col.sizeIsMax},
+				},
+			}
+		} else {
+			typ = &schema.UnsupportedType{T: col.typ}
+		}
+	}
+	return typ, nil
+}
+
+// indexes queries and appends the indexes of the given table.
+func (i *inspect) indexes(ctx context.Context, s *schema.Schema) error {
+	rows, err := i.querySchema(ctx, indexesQuery, s)
+	if err != nil {
+		return fmt.Errorf("spanner: querying schema %q indexes: %w", s.Name, err)
+	}
+	defer rows.Close()
+	if err := i.addIndexes(s, rows); err != nil {
+		return err
+	}
+	return rows.Err()
+}
+
+// addIndexes scans the rows and adds the indexes to the table.
+func (i *inspect) addIndexes(s *schema.Schema, rows *sql.Rows) error {
+	names := make(map[string]*schema.Index)
+	for rows.Next() {
+		var (
+			tableSchema                     sql.NullString
+			tableName, indexName, indexType string
+			parentTableName                 sql.NullString
+			isUnique, isNullFiltered        sql.NullBool
+			columnName                      sql.NullString
+			ordinalPosition                 sql.NullInt64
+			columnOrdering                  sql.NullString
+			isNullable                      sql.NullBool
+		)
+		if err := rows.Scan(
+			&tableSchema, &tableName, &indexName,
+			&indexType, &parentTableName, &isUnique,
+			&isNullFiltered, &columnName, &ordinalPosition,
+			&columnOrdering, &isNullable,
+		); err != nil {
+			return fmt.Errorf("spanner: scanning indexes for schema %q: %w", s.Name, err)
+		}
+		t, ok := s.Table(tableName)
+		if !ok {
+			return fmt.Errorf("table %q was not found in schema", tableName)
+		}
+		name := tableName + "_" + indexName
```

Comment:
I wonder why not using the same logic we used in other drivers:
https://github.com/ariga/atlas/blob/master/sql/mysql/inspect.go#L330-L339
https://github.com/ariga/atlas/blob/master/sql/postgres/inspect.go#L329-L334

Reply:
We want to apply Primary Keys during the index discovery process but all Primary Keys in Spanner are named `PRIMARY_KEY` so they need to be differentiated somehow. Just made a commit that hopefully clarifies this area a bit. Let me know what you think.